pkg/logql/filter.go

```diff
@@ -0,0 +1,287 @@
+package logql
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// LineFilter is a interface to filter log lines.
+type LineFilter interface {
+	Filter(line []byte) bool
+}
+
+// LineFilterFunc is a syntax sugar for creating line filter from a function
+type LineFilterFunc func(line []byte) bool
+
+func (f LineFilterFunc) Filter(line []byte) bool {
+	return f(line)
+}
+
+type notFilter struct {
+	LineFilter
+}
+
+func (n notFilter) Filter(line []byte) bool {
+	return !n.LineFilter.Filter(line)
+}
+
+// newNotFilter creates a new filter which matches only if the base filter doesn't match.
+// If the base filter is a `or` it will recursively simplify with `and` operations.
+func newNotFilter(base LineFilter) LineFilter {
+	// not(a|b) = not(a) and not(b)
+	if or, ok := base.(orFilter); ok {
```

Comment:
I'm not sure I understand why this is any better than the original form. Could you explain?

Reply:
cause a or close will always trigger both filter while an and is more strict and will stop if the first leg fails.

Reply:
Basically and are always faster as they stop as soon as one match is false.

Reply:
I thought `||` will short circuit too if the LHS is true? I don't think it matters that much, I think your simplification here is fine.

Reply:
Yes but we are talking about negation here. So if a doesn't match we will have to check b.

This is only true for negation ! This is because negation is waiting for the full result, if a is false, then b is executed which is not required when doing not.

Tricky to understand :)